lightning/src/routing/router.rs

```diff
@@ -1564,45 +1564,55 @@ fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters,
 	for path in route.paths.iter_mut() {
 		let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
 
-		// Choose the last publicly known node as the starting point for the random walk
-		if let Some(starting_hop) = path.iter().rev().find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.pubkey))) {
-			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.pubkey);
+		// Mark all nodes on the actual path as visited to avoid looping random walks.
+		let mut visited_nodes: HashSet<NodeId> = HashSet::new();
```

Comment:
I'd think it'd be faster to keep a single vec and just iterate it, but at a higer level, do we care about avoiding taking any node that we preiviously took in the path, or do we only care about not going back on the last three hops of the path, that'd let us even keep a fixed length array of like 3 and not hit the heap at all.

Reply:
> I'd think it'd be faster to keep a single vec and just iterate it, 

Hm, true, that might be faster since the routes are usually really not long. a26083ee24eb9c2371c3e53489b4ef099ef73a64 now utilizes a `Vec`.

> but at a higer level, do we care about avoiding taking any node that we preiviously took in the path, or do we only care about not going back on the last three hops of the path, that'd let us even keep a fixed length array of like 3 and not hit the heap at all.

Well, I'd argue that if we bother about looping shadow routes in the first place, we should do it 'properly' and avoid any loops. That said, if we don't want to hit the heap, we could also just use a fixed array with the maximum length of 20 to be on the safe side, no? (Which reminds me that I probably should limit the shadow route length so that the overall length of path + shadow route never exceeds 20, because otherwise it becomes implausible anyways).

Reply:
I realize I had the fixed pre-TLV limit in mind, but limiting the shadow route length in conjunction with https://github.com/lightningdevkit/rust-lightning/issues/1371 is probably still a good idea.

Reply:
> Well, I'd argue that if we bother about looping shadow routes in the first place, we should do it 'properly' and avoid any loops.

Right, I'm not sure if its really "properly" - if you're a node one hop from the end of a 5 hop route, the fact that the route appears to loop back over the first hop is actually a signal that that first hop node is not in the path, whereas if we don't consider looping over the first hop an issue such a node gets no signal. Its a bit of an awkward balancing act, though - allowing looping removes signal for nodes late in the path, but not allowing looping removes a signal for nodes early in the path.

> That said, if we don't want to hit the heap, we could also just use a fixed array with the maximum length of 20 to be on the safe side, no?

I mean its not a *huge* concern, just always nice to avoid it where possible - its never free and I'm generally scarred from overuse of the heap/allocations/reallocations in Bitcoin Core that causes severe heap fragmentation and memory usage bloat in excess of what it really needs...something Core has never been in a position to materially improve :(.

Reply:
Yea, I think we're on the same page. I'd be happy to see you go to only the last three hops, I'll leave it up to you exactly how you want to proceed, though, wanted to get on the same page first.

Reply:
With https://github.com/lightningdevkit/rust-lightning/pull/1360/commits/85dbc5ddf03b386c75cb587c0ae0c6c32a1ca704, I switched to a fixed-length array that remembers only the three most recent nodes, which is initialized from the path suffix if possible.